README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fec16 <img src="data-raw/Sticker/hex_fec16.png" align="right" height=140/>

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://r-pkg.org/badges/version/fec16)](https://CRAN.R-project.org/package=fec16)
[![Travis-CI Build
Status](https://travis-ci.org/baumer-lab/fec16.svg?branch=master)](https://travis-ci.org/baumer-lab/fec16)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/fec16)](https://www.r-pkg.org:443/pkg/fec16)
<!-- badges: end -->

**fec16** contains data from the [Federal Election Commission
(FEC)](https://www.fec.gov/) website pertaining to candidates,
committees, results, contributions from committees and individuals, and
other financial [data for the United States 2015-2016 election
cycle](https://www.fec.gov/data/browse-data/?tab=bulk-data).
Additionally, for the datasets that are included as samples, the package
includes functions that import the full versions.

## Installation

Get the latest released version from CRAN:

``` r
install.packages("fec16")
```

Or the development version from GitHub:

``` r
# If you haven't installed the remotes package yet, do so:
# install.packages("remotes")
remotes::install_github("baumer-lab/fec16")
```

``` r
# Load package
library(fec16)
```

## Datasets Included

### Full Datasets

  - `candidates`: candidates registered with the FEC during the
    2015-2016 election cycle
  - `committees`: committees registered with the FEC during the
    2015-2016 election cycle
  - `campaigns`: the House/Senate current campaigns
  - `results_house`: the House results of the 2016 general election
  - `results_senate`: the Senate results of the 2016 general election
  - `results_president`: the final results of the 2016 general election
  - `pac`: Political Action Committee (PAC) and party summary financial
    information
  - `states`: geographical information about the 50 states

### Sample Datasets (with 1000 random rows each)

  - `individuals`: individual contributions to candidates/committees
    during the 2016 election cycle
  - `contributions`: candidates and their contributions from committees
    during the 2016 election cycle
  - `expenditures`: the operating expenditures
  - `transactions`: transactions between committees

## Functions Included

The following functions retrieve the entire datasets for the sampled
ones listed above. The size of the raw file that is downloaded by
calling each function is given for reference. All functions have an
argument `n_max` which defaults to the entire dataset but the user can
specify the max length of the dataset to be loaded via this argument.

  - `read_all_individuals()` ~ 1.45GB
  - `read_all_contributions()` ~ 15.4MB
  - `read_all_expenditures()` ~ 52.1MB
  - `read_all_transactions()` ~ 79.2MB

## How is the data relational?

The headers of each table show the dataset name. The underlined
variables are **primary keys** while all the others are **foreign
keys**. The arrows show how the datasets are connected.

<img src="inst/fec16-dm.jpeg" align="center"/>

The diagram is built using the `dm` R package. The code can be found in
`data-raw/dm.R`.

## Examples

### Data Wrangling

`fec16` can be used to summarize data in order see how many candidates
are running for elections (in all offices) for the two major parties:

``` r
library(dplyr)

data <- candidates %>%
  filter(cand_pty_affiliation %in% c("REP", "DEM")) %>%
  group_by(cand_pty_affiliation) %>%
  summarize(size = n())

data
#> # A tibble: 2 x 2
#>   cand_pty_affiliation  size
#>   <chr>                <int>
#> 1 DEM                   1270
#> 2 REP                   1495
```

### Data Visualization

We can visualize the above data:

``` r
library(ggplot2)

ggplot(data, aes(x = cand_pty_affiliation, y = size, fill = cand_pty_affiliation)) +
  geom_col() +
  labs(
    title = "Number of Candidates Affiliated with the Two Major Parties",
    x = "Party", y = "Count", fill = "Candidate Party Affiliation"
  )
```

<img src="man/figures/README-party-plot-1.png" width="100%" />

## See Also

If you are interested in political data, check out the following related
packages:

  - [`politicaldata`](https://github.com/elliottmorris/politicaldata/)
  - [`ProPublicaR`](https://github.com/dietrichson/ProPublicaR)

## Contributors

  - [Marium Tapal](https://github.com/mariumtapal)
  - [Irene Ryan](https://github.com/ireneryan)
  - [Rana Gahwagy](https://github.com/ranawg)
  - [Benjamin S. Baumer](https://github.com/beanumber)


NEWS.md
# fec16 0.1.3

* Added `states` data frame
* Solved discrepancy in the `party` variable in the `results_house` and `results_senate` datasets
* Updated README to add `dm` object and related R packages

# fec16 0.1.2 (2020-10-16)

Resubmission

* Fixed issues with dependencies and imports

# fec16 0.1.1 (2020-08-18)

Resubmission

* Edited internal vignette code for faster compilation
* Fixed the 4 possible invalid URLs 
>>>>>>> f8d6a3e3ca36204bba0dd45a79d1658c40ae2b9e

# fec16 0.1.0 (2020-07-10)

First release for CRAN

* Added data pertaining to candidates, committees, results, contributions from committees and individuals, and other financial data for the United States 2015-2016 election cycle. 
* Included functions that import the full versions of the datasets that are included as samples.
* Included package vignette


R/fec16-package.R
#' fec16: A Data Package for the 2016 Elections
#'
#' The fec16 package provides the following 9 datasets from the Federal Election Commission's 2015-2016 election cycle:
#'
#' @section Datasets:
#' \if{html}{\figure{hex_fec16.png}{options: align = "right" width="45\%" alt="Figure: hex_fec16.png"}}
#' \itemize{
#'   \item Candidates: [candidates]
#'   \item Committees: [committees]
#'   \item House Results: [results_house]
#'   \item Senate Results: [results_senate]
#'   \item Presidential Results: [results_president]
#'   \item Individual Contributions: [individuals]
#'   \item Committee Contributions: [contributions]
#'   \item House/Senate Current Campaigns: [campaigns]
#'   \item Operating Expenditures: [expenditures]
#'   \item Transactions From One Committee To Another: [transactions]
#'   \item Political Action Committee (PAC) and Party Summary Financial Information: [pac]}
#'
#' The entire [candidates], [committees], [campaigns], [results_house], [results_senate], [results_president], and [pac] datasets and samples containing 1000 entries of the other datasets are available directly in the package.
#' For the datasets containing samples, full versions can be retrieved via the following functions:
#' \itemize{
#'   \item Individual Contributions: [read_all_individuals]
#'   \item Committee Contributions: [read_all_contributions]
#'   \item Operating Expenditures: [read_all_expenditures]
#'   \item Transactions From One Committee To Another: [read_all_transactions]
#' }
#'
#' In addition, `fec16` provides the [states] dataset giving geographical information about the 50 states.
#'
#' @section Authors:
#' \itemize{
#'   \item Marium Tapal: \url{https://github.com/mariumtapal}
#'   \item Irene Ryan: \url{https://github.com/ireneryan}
#'   \item Rana Gahwagy: \url{https://github.com/ranawg}
#'   \item Ben Baumer: \url{https://github.com/beanumber}
#' }
#'
#' @docType package
#' @name fec16
NULL


R/fn_individuals.R
#' Individual Contributions Master metadata
#'
#' \code{read_all_individuals} returns a dataframe about transactions from individuals to candidates/committees
#'
#' @param n_max Integer specifying the max amount of entries in the dataset. Defaults to the possible maximum.
#' @param verbose A progress bar is shown if R is running interactively. Defaults to `interactive()`.
#' @return The entire dataframe. More information about variables is at `?individuals`.
#' @examples
#' \donttest{read_all_individuals()}
#' \donttest{read_all_individuals(n_max = 250)}
#' @import dplyr
#' @import readr
#' @export

read_all_individuals <- function(n_max = Inf, verbose = interactive()) {
  if (!verbose) {
    invisible(utils::capture.output(
      dir <- usethis::use_zip(
        "https://www.fec.gov/files/bulk-downloads/2016/indiv16.zip",
        destdir = tempdir(), cleanup = TRUE
      )
    ))
  } else {
    dir <- usethis::use_zip(
      "https://www.fec.gov/files/bulk-downloads/2016/indiv16.zip",
      destdir = tempdir(), cleanup = TRUE
    )
  }

  indiv_path <- fs::path(dir, "itcont.txt")

  indiv_names <- read_csv("https://www.fec.gov/files/bulk-downloads/data_dictionaries/indiv_header_file.csv") %>%
    names() %>%
    tolower()

  individuals_all <- vroom::vroom(
    indiv_path,
    n_max = n_max,
    col_names = indiv_names,
    col_types = cols(
      zip_code = col_character(),
      other_id = col_character(),
      memo_cd = col_character(),
      memo_text = col_character(),
      sub_id = col_character(),
      transaction_tp = col_character()
    ),
    delim = "|"
  ) %>%
    select(-c(image_num, sub_id, memo_text, memo_cd, file_num)) %>%
    mutate(
      transaction_dt = lubridate::mdy(transaction_dt)
    )
  return(individuals_all)
}


R/fn_expenditures.R
#' Operating Expenditures
#'
#' \code{read_all_expenditures} returns a dataframe about operating expenditure data
#'
#' @param n_max Integer specifying the max amount of entries in the dataset. Defaults to the possible maximum.
#' @param verbose A progress bar is shown if R is running interactively. Defaults to `interactive()`.
#' @return The entire dataframe. More information about variables is at `?expenditures`.
#' @examples
#' \donttest{read_all_expenditures()}
#' \donttest{read_all_expenditures(n_max = 250)}
#' @import dplyr
#' @import readr
#' @export

read_all_expenditures <- function(n_max = Inf, verbose = interactive()) {
  if (!verbose) {
    invisible(utils::capture.output(
      dir <- usethis::use_zip(
        "https://www.fec.gov/files/bulk-downloads/2016/oppexp16.zip",
        destdir = tempdir(), cleanup = TRUE
      )
    ))
  } else {
    dir <- usethis::use_zip(
      "https://www.fec.gov/files/bulk-downloads/2016/oppexp16.zip",
      destdir = tempdir(), cleanup = TRUE
    )
  }

  oppexp_path <- fs::path(dir, "oppexp.txt")

  oppexp_names <- read_csv("https://www.fec.gov/files/bulk-downloads/data_dictionaries/oppexp_header_file.csv") %>%
    names() %>%
    tolower()

  oppexp_all <- vroom::vroom(
    oppexp_path,
    n_max = n_max,
    col_names = oppexp_names,
    col_types = cols(
      zip_code = col_character(),
      tran_id = col_character(),
      memo_cd = col_character(),
      memo_text = col_character(),
      sub_id = col_character()
    ),
    delim = "|"
  )

  expenditures_all <- oppexp_all %>%
    select(-c(image_num, sub_id, memo_text, memo_cd, file_num, X26)) %>%
    mutate(
      transaction_dt = lubridate::mdy(transaction_dt)
    )

  return(expenditures_all)
}


R/fn_transactions.R
#' Any Transaction From One Committee To Another
#'
#' \code{read_all_transactions} returns a dataframe about transaction data
#'
#' @param n_max Integer specifying the max amount of entries in the dataset. Defaults to the possible maximum.
#' @param verbose A progress bar is shown if R is running interactively. Defaults to `interactive()`.
#' @return The entire dataframe. More information about variables is at `?transactions`.
#' @examples
#' \donttest{read_all_transactions()}
#' \donttest{read_all_transactions(n_max = 250)}
#' @import dplyr
#' @import readr
#' @export

read_all_transactions <- function(n_max = Inf, verbose = interactive()) {
  if (!verbose) {
    invisible(utils::capture.output(
      dir <- usethis::use_zip(
        "https://www.fec.gov/files/bulk-downloads/2016/oth16.zip",
        destdir = tempdir(), cleanup = TRUE
      )
    ))
  } else {
    dir <- usethis::use_zip(
      "https://www.fec.gov/files/bulk-downloads/2016/oth16.zip",
      destdir = tempdir(), cleanup = TRUE
    )
  }

  transactions_path <- fs::path(dir, "itoth.txt")

  transactions_names <- read_csv("https://www.fec.gov/files/bulk-downloads/data_dictionaries/oth_header_file.csv") %>%
    names() %>%
    tolower()

  transactions_all <- vroom::vroom(
    transactions_path,
    n_max = n_max,
    col_names = transactions_names,
    col_types = cols(
      zip_code = col_character(),
      other_id = col_character(),
      memo_cd = col_character(),
      memo_text = col_character(),
      sub_id = col_character()
    ),
    delim = "|"
  )

  transactions <- transactions_all %>%
    select(-c(image_num, sub_id, memo_text, memo_cd, file_num)) %>%
    mutate(
      transaction_dt = lubridate::mdy(transaction_dt)
    )
  return(transactions)
}


R/fn_contributions.R
#' Committee contributions metadata
#'
#' \code{read_all_contributions} returns a dataframe about contribution data
#'
#' @param n_max Integer specifying the max amount of entries in the dataset. Defaults to the possible maximum.
#' @param verbose A progress bar is shown if R is running interactively. Defaults to `interactive()`.
#' @return The entire dataframe. More information about variables is at `?contributions`.
#' @examples
#' \donttest{read_all_contributions()}
#' \donttest{read_all_contributions(n_max = 250)}
#' @import dplyr
#' @import readr
#' @export

read_all_contributions <- function(n_max = Inf, verbose = interactive()) {
  if (!verbose) {
    invisible(utils::capture.output(
      dir <- usethis::use_zip(
        "https://www.fec.gov/files/bulk-downloads/2016/pas216.zip",
        destdir = tempdir(), cleanup = TRUE
      )
    ))
  } else {
    dir <- usethis::use_zip(
      "https://www.fec.gov/files/bulk-downloads/2016/pas216.zip",
      destdir = tempdir(), cleanup = TRUE
    )
  }

  file_path <- fs::path(dir, "itpas2.txt")

  exp_names <- read_csv("https://www.fec.gov/files/bulk-downloads/data_dictionaries/pas2_header_file.csv") %>%
    names() %>%
    tolower()

  contributions_all <- vroom::vroom(
    file_path,
    col_names = exp_names,
    col_types = cols(
      employer = col_character(),
      occupation = col_character()
    ),
    n_max = n_max,
    delim = "|"
  ) %>%
    select(-c(employer, occupation, image_num, memo_cd, memo_text, sub_id, file_num)) %>%
    mutate(
      transaction_dt = lubridate::mdy(transaction_dt)
    )
  return(contributions_all)
}


R/data.R
#' Candidates Master metadata
#'
#' @description Basic information for each candidate.
#'
#' @format A data frame with 4648 rows and 15 columns, providing information
#' for each candidate:
#' \describe{
#'   \item{cand_id}{A 9-character alpha-numeric code assigned to a candidate by the FEC, which remains
#'   the same across election cycles if running for the same office}
#'   \item{cand_name}{Candidate name}
#'   \item{cand_pty_affiliation}{The political party affiliation reported by the candidate}
#'   \item{cand_election_yr}{Candidate's Election Year}
#'   \item{cand_office_st}{Candidate's state}
#'   \item{cand_office}{Candidate office with designation: H = House, P = President, S = Senate}
#'   \item{cand_office_district}{Congressional district number: Congressional at-large 00, Senate 00, Presidential 00}
#'   \item{cand_ici}{Incumbent challenger status with designation: C = Challenger, I = Incumbent, O = Open Seat}
#'   \item{cand_status}{Candidate status with designation: C = Statutory candidate, F = Statutory candidate for future election,
#'   N = Not yet a statutory candidate, P = Statutory candidate in prior cycle}
#'   \item{cand_pcc}{ID of the candidate's principal campaign committee assigned by FEC}
#'   \item{cand_st1}{Candidate's Mailing address - street 1}
#'   \item{cand_st2}{Candidate's Mailing address - street 2}
#'   \item{cand_city}{Candidate's Mailing address - city}
#'   \item{cand_st}{Candidate's Mailing address - state}
#'   \item{cand_zip}{Candidate's Mailing address - ZIP code}
#' }
#' @source <https://www.fec.gov/campaign-finance-data/candidate-master-file-description/>, updated 2020-03-02
"candidates"

#' Committee contributions metadata
#'
#' @description Provides data pertaining to each candidate and their contributions from committees during the 2016 general election.
#'
#' @format A data frame with 1000 rows and 15 columns with information about candidate committee contributions.
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{amndt_ind}{Amendment indicator: Indicates if the report being filed is new (N), an amendment (A) to a previous
#'   report or a termination (T) report}
#'   \item{rpt_tp}{Indicates the type of report filed, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/report-type-code-descriptions/}}
#'   \item{transaction_pgi}{This code indicates the election for which the contribution was made.
#'   EYYYY (election Primary, General, Other plus election year)}
#'   \item{transaction_tp}{Transaction types, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/transaction-type-code-descriptions/}}
#'   \item{entity_tp}{Entity Type:
#'
#'   CAN = Candidate
#'
#'   CCM = Candidate Committee
#'
#'   COM = Committee
#'
#'   IND = Individual (a person)
#'
#'   ORG = Organization (not a committee and not a person)
#'
#'   PAC = Political Action Committee
#'
#'   PTY = Party Organization}
#'   \item{name}{Contributor/lender/transfer Name}
#'   \item{city}{City}
#'   \item{state}{State}
#'   \item{zip_code}{ZIP Code}
#'   \item{transaction_dt}{Transaction date (YYYY-MM-DD)}
#'   \item{transaction_amt}{Transaction Amount}
#'   \item{other_id}{For contributions from individuals this column is null.
#'   For contributions from candidates or other committees this column will contain that contributor's FEC ID}
#'   \item{cand_id}{A 9-character alpha-numeric code assigned to a candidate by the FEC, which remains
#'   the same across election cycles if running for the same office}
#'   \item{tran_id}{Only for Electronic Filings. A unique identifier associated with each itemization or transaction
#'   appearing in an FEC electronic file. A transaction ID is unique for a specific committee for a specific report.}
#' }
#' @source <https://www.fec.gov/campaign-finance-data/contributions-committees-candidates-file-description/>, Updated 2020-03-02
"contributions"

#' Committees metadata
#'
#' @description Provides data pertaining to each committee registered with the Federal Election Commission.
#'
#' @format A data frame with 17654 rows and 15 columns with information about FEC committees.
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{cmte_nm}{Name of the committee}
#'   \item{tres_nm}{The officially registered treasurer for the committee}
#'   \item{cmte_st1}{Street 1}
#'   \item{cmte_st2}{Street 2}
#'   \item{cmte_city}{City}
#'   \item{cmte_st}{State}
#'   \item{cmte_zip}{ZIP Code}
#'   \item{cmte_dsgn}{A = Authorized by a candidate
#'
#' B = Lobbyist/Registrant PAC
#'
#' D = Leadership PAC
#'
#' J = Joint fundraiser
#'
#' P = Principal campaign committee of a candidate
#'
#' U = Unauthorized}
#'   \item{cmte_tp}{Committee Type, listed here: \url{https://www.fec.gov/campaign-finance-data/committee-type-code-descriptions/}}
#'   \item{cmte_pty_affiliation}{Committee Party, listed here: \url{https://www.fec.gov/campaign-finance-data/party-code-descriptions/}}
#'   \item{cmte_filing_freq}{Filing frequency: A = Administratively terminated,
#'
#'   D = Debt
#'
#'   M = Monthly filer
#'
#'   Q = Quarterly filer
#'
#'   T = Terminated
#'
#'   W = Waived}
#'   \item{org_tp}{Interest group category:
#'
#'   C = Corporation
#'
#'   L = Labor organization
#'
#'   M = Membership organization
#'
#'   T = Trade association
#'
#'   V = Cooperative
#'
#'   W = Corporation without capital stock}
#'   \item{connected_org_nm}{Connected organization's name}
#'   \item{cand_id}{Candidate identification:When a committee has a committee type designation of H, S, or P,
#'   the candidate's identification number will be entered in this field.}
#' }
#' @source <https://www.fec.gov/campaign-finance-data/committee-master-file-description/>, updated 2020-03-02
"committees"

#' Individual Contributions Master metadata
#'
#' @description Information for a simple random sample of transactions from
#' individuals to candidates/committees.
#'
#' @format A data frame with 1000 rows and 16 columns providing information for each transaction:
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{amndt_ind}{Amendment indicator: Indicates if the report being filed is new (N), an amendment (A) to a previous
#'   report or a termination (T) report}
#'   \item{rpt_tp}{Indicates the type of report filed, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/report-type-code-descriptions/}}
#'   \item{transaction_pgi}{This code indicates the election for which the contribution was made.
#'   EYYYY (election Primary, General, Other plus election year)}
#'   \item{transaction_tp}{Transaction types, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/transaction-type-code-descriptions/}}
#'   \item{entity_tp}{Entity Type:
#'
#'   CAN = Candidate
#'
#'   CCM = Candidate Committee
#'
#'   COM = Committee
#'
#'   IND = Individual (a person)
#'
#'   ORG = Organization (not a committee and not a person)
#'
#'   PAC = Political Action Committee
#'
#'   PTY = Party Organization}
#'   \item{name}{Contributor/lender/Transfer Name}
#'   \item{city}{City}
#'   \item{state}{State}
#'   \item{zip_code}{ZIP Code}
#'   \item{employer}{Employer}
#'   \item{occupation}{Occupation}
#'   \item{transaction_dt}{Transaction date (YYYY-MM-DD)}
#'   \item{transaction_amt}{Transaction Amount}
#'   \item{other_id}{For contributions from individuals this column is null.
#'   For contributions from candidates or other committees this column will contain that contributor's FEC ID.}
#'   \item{tran_id}{Only for Electronic Filings. A unique identifier associated with each itemization or transaction
#'   appearing in an FEC electronic file. A transaction ID is unique for a specific committee for a specific report.}
#' }
#' @source <https://www.fec.gov/campaign-finance-data/contributions-individuals-file-description/>, updated 2020-03-02
"individuals"

#' General Federal Election Results
#'
#' @description 2016 House General Election Results
#'
#' @format 3 data frames with information about the House, Senate, and Presidential Election results:
#' \describe{
#'   \item{state}{The state of the votes}
#'   \item{district_id}{District ID}
#'   \item{cand_id}{A code assigned to a candidate by the FEC}
#'   \item{incumbent}{A logical variable: TRUE if incumbent, FALSE if not}
#'   \item{party}{Party of the candidate}
#'   \item{primary_votes}{Number of votes in the primary election}
#'   \item{primary_percent}{Percentage of votes in the primary election}
#'   \item{runoff_votes}{Number of runoff votes}
#'   \item{runoff_percent}{Percentage of runoff votes}
#'   \item{general_votes}{Number of votes in the general election}
#'   \item{general_percent}{Percentage of votes in the general election}
#'   \item{won}{A logical variable: TRUE if candidate won, FALSE if not}
#'   \item{footnotes}{Footnotes}
#' }
#' @source <https://transition.fec.gov/general/FederalElections2016.shtml>, updated 2020-03-02
"results_house"

#' @rdname results_house
#' @description 2016 Senate General Election Results
"results_senate"

#' @rdname results_house
#' @description 2016 Presidential General Election Results
"results_president"

#' House/Senate Current Campaigns
#'
#' @description Summary financial information for each campaign.
#'
#' @format A data frame with 1898 rows and 25 columns, providing financial
#' information for each House/Senate campaign:
#' \describe{
#'   \item{cand_id}{A code assigned to a candidate by the FEC}
#'   \item{cand_name}{Candidate name}
#'   \item{cand_ici}{Incumbent challenger status with designation:
#'    C = Challenger, I = Incumbent, O = Open Seat}
#'   \item{pty_cd}{Party code}
#'   \item{cand_pty_affiliation}{The political party affiliation reported by the candidate}
#'   \item{ttl_receipts}{Total receipts}
#'   \item{trans_from_auth}{Transfers from authorized committees}
#'   \item{ttl_disb}{Total disbursements}
#'   \item{trans_to_auth}{Transfers to authorized committees}
#'   \item{coh_bop}{Beginning cash}
#'   \item{coh_cop}{Ending cash}
#'   \item{cand_contrib}{Contributions from candidate}
#'   \item{cand_loans}{Loans from candidate}
#'   \item{other_loans}{Other loans}
#'   \item{cand_loan_repay}{Candidate loan repayments}
#'   \item{other_loan_repay}{Other loan repayments}
#'   \item{debts_owed_by}{Debts owed by}
#'   \item{ttl_indiv_contrib}{Total individual contribution}
#'   \item{cand_office_st}{Candidate's state}
#'   \item{cand_office_district}{Candidate's district}
#'   \item{other_pol_cmte_contrib}{Contributions from other political committees}
#'   \item{pol_pty_contrib}{Contributions from party committees}
#'   \item{cvg_end_dt}{Coverage end date (YYYY-MM-DD)}
#'   \item{indiv_refunds}{Refunds to individuals}
#'   \item{cmte_refunds}{Refunds to committees}
#' }
#' @source <https://transition.fec.gov/general/FederalElections2016.shtml>, updated 2020-03-02
"campaigns"

#' Operating Expenditures
#'
#' @description Operating expenditure information for a simple random sample
#' of 1000 entries
#'
#' @format A data frame with 1000 rows and 20 columns:
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{amndt_ind}{Amendment indicator: Indicates if the report being filed is new (N), an amendment (A) to a previous
#'   report or a termination (T) report}
#'   \item{rpt_yr}{Report year}
#'   \item{rpt_tp}{Indicates the type of report filed, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/report-type-code-descriptions/}}
#'   \item{line_num}{Indicates FEC form line number}
#'   \item{form_tp_cd}{Indicates FEC form type}
#'   \item{sched_tp_cd}{Schedule type: Schedule B - Itemized disbursements}
#'   \item{name}{Contributor/lender/Transfer Name}
#'   \item{city}{City}
#'   \item{state}{State}
#'   \item{zip_code}{ZIP Code}
#'   \item{transaction_dt}{Transaction date (YYYY-MM-DD)}
#'   \item{transaction_amt}{Transaction Amount}
#'   \item{transaction_pgi}{Primary general indicator}
#'   \item{purpose}{Purpose}
#'   \item{category}{Disbursement category code: 001-012 and 101-107}
#'   \item{category_desc}{Disbursement Category Code Description,
#'   listed here: \url{https://www.fec.gov/campaign-finance-data/disbursement-category-code-descriptions/}}
#'   \item{entity_tp}{Entity Type:
#'
#'   CAN = Candidate
#'
#'   CCM = Candidate Committee
#'
#'   COM = Committee
#'
#'   IND = Individual (a person)
#'
#'   ORG = Organization (not a committee and not a person)
#'
#'   PAC = Political Action Committee
#'
#'   PTY = Party Organization}
#'
#'   \item{tran_id}{Only for Electronic Filings. A unique identifier associated with each itemization or transaction
#'   appearing in an FEC electronic file. A transaction ID is unique for a specific committee for a specific report.}
#'
#'   \item{back_ref_tran_id}{Back reference transaction ID: Used to associate one transaction with another transaction in the same report
#'   (using file number, transaction ID and back reference transaction ID).
#'   For example, a credit card payment and the sub-itemization of specific purchases.
#'   The back reference transaction ID of the specific purchases will equal the transaction ID of the payment to the credit card company.}
#'   }
#' @source <https://www.fec.gov/campaign-finance-data/operating-expenditures-file-description/>, updated 2020-03-02
"expenditures"

#' Any Transaction From One Committee To Another
#'
#' @description Transaction information between committees
#'
#' @format A data frame with 1000 rows and 16 columns:
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{amndt_ind}{Amendment indicator: Indicates if the report being filed is new (N), an amendment (A) to a previous
#'   report or a termination (T) report}
#'   \item{rpt_tp}{Indicates the type of report filed, listed here:
#'   \url{https://www.fec.gov/campaign-finance-data/report-type-code-descriptions/}}
#'   \item{transaction_pgi}{Primary general indicator:This code indicates the election for which the contribution was made.
#'   EYYYY (election Primary, General, Other plus election year)}
#'   \item{transaction_tp}{Transaction type, listed here: \url{https://www.fec.gov/campaign-finance-data/transaction-type-code-descriptions}}
#'   \item{entity_tp}{Entity Type:
#'
#'   CAN = Candidate
#'
#'   CCM = Candidate Committee
#'
#'   COM = Committee
#'
#'   IND = Individual (a person)
#'
#'   ORG = Organization (not a committee and not a person)
#'
#'   PAC = Political Action Committee
#'
#'   PTY = Party Organization}
#'   \item{name}{Contributor/lender/Transfer Name}
#'   \item{city}{City}
#'   \item{state}{State}
#'   \item{zip_code}{ZIP Code}
#'   \item{employer}{Employer}
#'   \item{occupation}{Occupation}
#'   \item{transaction_dt}{Transaction date (YYYY-MM-DD)}
#'   \item{transaction_amt}{Transaction Amount}
#'   \item{other_id}{For contributions from individuals this column is null.
#'   For contributions from candidates or other committees this column will contain that contributor's FEC ID}
#'   \item{tran_id}{Only for Electronic Filings. A unique identifier associated with each itemization or transaction
#'   appearing in an FEC electronic file. A transaction ID is unique for a specific committee for a specific report.}
#'   }
#' @source <https://www.fec.gov/campaign-finance-data/any-transaction-one-committee-another-file-description/>, updated 2020-03-02
"transactions"

#' Political Action Committee (PAC) and Party Summary Financial Information
#'
#' @description PAC and party summary financial information
#'
#' @format A data frame with 12049 rows and 27 columns:
#' \describe{
#'   \item{cmte_id}{A 9-character alpha-numeric code assigned to a committee by the Federal Election Commission}
#'   \item{cmte_nm}{Name of the committee}
#'   \item{cmte_tp}{Committee Type, listed here: \url{https://www.fec.gov/campaign-finance-data/committee-type-code-descriptions/}}
#'   \item{cmte_dsgn}{A = Authorized by a candidate
#'
#' B = Lobbyist/Registrant PAC
#'
#' D = Leadership PAC
#'
#' J = Joint fundraiser
#'
#' P = Principal campaign committee of a candidate
#'
#' U = Unauthorized}
#'
#'   \item{cmte_filing_freq}{Filing frequency: A = Administratively terminated,
#'
#'   D = Debt
#'
#'   M = Monthly filer
#'
#'   Q = Quarterly filer
#'
#'   T = Terminated
#'
#'   W = Waived}
#'
#'   \item{ttl_receipts}{Total receipts}
#'   \item{trans_from_aff}{Transfers from affiliates}
#'   \item{indv_contrib}{Contributions from individuals}
#'   \item{other_pol_cmte_contrib}{Contributions from other political committees}
#'   \item{cand_contrib}{Contributions from candidate}
#'   \item{cand_loans}{Candidate loans}
#'   \item{ttl_loans_received}{Total loans received}
#'   \item{ttl_disb}{Total disbursements}
#'   \item{tranf_to_aff}{Transfers to affiliates}
#'   \item{indv_refunds}{Refunds to individuals}
#'   \item{other_pol_cmte_refunds}{Refunds to other political committees}
#'   \item{cand_loan_repay}{Candidate loan repayments}
#'   \item{loan_repay}{Loan repayments}
#'   \item{coh_bop}{Cash beginning of period}
#'   \item{coh_cop}{Cash close of period}
#'   \item{debts_owed_by}{Debts owed by}
#'   \item{nonfed_trans_received}{Non-federal transfers received}
#'   \item{contrib_to_other_cmte}{Contributions to other committees}
#'   \item{ind_exp}{Independent expenditures}
#'   \item{pty_coord_exp}{Party coordinated expenditures}
#'   \item{nonfed_share_exp}{Non-federal share expenditures}
#'   \item{cvg_end_dt}{Coverage end date (YYYY-MM-DD)}
#'   }
#' @source <https://www.fec.gov/campaign-finance-data/pac-and-party-summary-file-description/>, updated 2020-03-02
"pac"

#' US States lookup table
#' @docType data
#' @format A data frame with 50 rows and 4 columns:
#' \describe{
#'   \item{state}{two-character postal abbreviation}
#'   \item{state_name}{the name of the state}
#'   \item{state_region}{coarse region of the country that the state belongs to}
#'   \item{state_division}{finer region of the country that the state belongs to}
#'   }
#' @source \code{\link[datasets]{state}}
"states"
